Personalization

austinpaine opened the pull request:
Tested with uberjar successfully.  Combined implementation of Chad's login page and user management, and my layout saving and loading.

Add button is hidden until user logs in, at which point any widgets associated with their name are loaded to the page.  User can logout and screen clears, when different user logs in page loads their widgets.  

Test it out yourself!  Make a couple different users and add some widgets to each, log in and out and in with a different user.  Let us know of any bugs

Commit:
test layout

Commit:
added DB table creation

Commit:
created new event to load a pre-built widget vector

Commit:
Created db.core comments

Commit:
Created db.core comments missed some files

Commit:
merged nrepl and color picker in from master

Commit:
got db table working and inserting.  MODAL merge fix in this commit

Commit:
made options inserting work as intended, created route and html handler

Commit:
Almost have it working and loading in from the DB, why did assoc suddenly not work??

cawasser committed:
reloading widgets partly works (widget sort of gets drawn) but incompatible keywords between SQLite and re-frame are still to be worked (:data-source vs :data_source and :data-grid vs :data_grid)

cawasser committed:
reloading widgets partly works (widget sort of gets drawn) but incompatible keywords between SQLite and re-frame are still to be worked (:data-source vs :data_source and :data-grid vs :data_grid)

cawasser committed:
set :key back to :id and used clojure.set/rename-keys to fix up all the key name changes with SQLite

Commit:
So close to working but the colons are being stripped off keys and they are strings when they get to the event in vanilla.core

Commit:
IT WORKS, layout loads in from the DB.  Cleaned up all tracing comments

Commit:
working the POST to save the layout to the db.  Confused.  in progress.

Commit:
mid-work on POST after chris talk, put things in the right places

Commit:
Post compiling but page errors

Commit:
Still trying to figure out errors, committing so i can roll back to older commits

Commit:
Merge branch 'master' into layout-saving

Commit:
merged with master but onLayoutChange didnt seem to merge right? removed the messy if in there to match master

Commit:
had to move the save-layout event into events namespace for some reason, but the page renders!

cawasser committed:
getting simple ajax POST call over to the server, but the data is not in the correct place in the request

Commit:
merged with chris updates

chadarm committed:
Added sql queries and hugsql code to create and access new users for login.

Commit:
IT WORKS!  layout post has all the data, just needs formatting now.  Added ring.json middleware

chadarm committed:
Added uri handlers for accessing user database.

chadarm committed:
Added ui to show a login button with attached modal. Login forms not there yet.

chadarm committed:
Added interactive email and password forms. Not hooked up to AJAX calls.

Commit:
Things are going to get real messy.  losing my mind trying to format the incoming data structure

chadarm committed:
Have Post calls saving info to db. Have GET calls retrieving information. Now need to add verification that username and password match db records.

chadarm committed:
added sql and tests in db core

chadarm committed:
Login verification up and running and is set up to send a dispatch call based on whether the login was valid. Dispatch calls are not set up yet.

chadarm committed:
Login verification is now up and running, previously was not. Now login sets the current user value in the app-db.

Commit:
LOADING WORKS!!!! need to fix next-id. committing with all comments for posterity before cleanup

Commit:
ITS DONE! Had to move the main save-layout ajax/POST into update-layout because it had to happen after the reduce-layouts call

Commit:
Cleanup

Commit:
merged with master

Commit:
adding an empty but created layout-table to the db

Commit:
Various cleanup and review comments.  New snapshot version, updated vanilla_db with new empty layout-table

Commit:
Updated Services table from VARCHAR to TEXT and re-created table in the DB

Commit:
Created db initialization function for repl

Commit:
Fixed next-id=NaN bug on initial page load when layout table is empty

Commit:
Refactored events namespace out into each events caller namespace

Commit:
added/tested a delete widget route to update database when user clicks x.

chadarm committed:
Login and logout button fully operational, modal pop up does not work on log out.

Commit:
Added a route for updating a single widget when the color is changed

Commit:
forgot a comma

chadarm committed:
Add login button to main dashboard.

chadarm committed:
Cleaned up some ui alignments.

chadarm committed:
Added in widget grid, project should be running normally now with the addition of the login feature.

chadarm committed:
merged in login branch.

Commit:
popups are hard

Commit:
merged latest layout-saving branch in

Commit:
Revert "popups are hard", didnt intend to merge this yet

This reverts commit 24dc5d25

Commit:
Moved login events to login-modal, added real username to widget-add

Commit:
fixed dbcore initializing function to not get called.

chadarm committed:
Toggle between login button and logout/add-widget button.

chadarm committed:
cleaned up and added constructive comments to things. Seems to work consistently.

Commit:
added functionality to load user specific widgets on login. fixed parinfer login bug

Commit:
screen clears of widgets on logout

Commit:
implemented unique id (UUID) generation for fully unique widget keys across various users

Commit:
deleted deprecated events.cljs ns, removed next-id db event since its handled by layout loading now

Commit:
forgot to cleanup events namespace requires in other ns's

Commit:
cleanup

Commit:
Merge branch 'master' into personalization

bfhub reviewed (commented):
recently three widgets have been added to Vanilla project, Health and Status, Current Time, and Data Table.
I logged in, and added 4 widgets, one from Spectrum Traces, and the rest, all the new widgets.
I logged out, and logged in again, the only widget that was displayed was the spectrum one.
does anyone see the same behavior? please let me know if I am missing something.

austinpaine commented:
@bfhub ooooh yep, i see bad behavior trying those new widgets.  I hadnt tested it with those yet.  Data table completely breaks the page, and the other two dont get added to the db because i need to add the :build-fn key value to the insert.  Thank you for testing it and catching that!  I'll look into fixing it tomorrow

cawasser commented on README.md:
```diff
@@ -23,7 +23,7 @@ This single-page application (SPA) provides a number of graphical widgets based
 - [ ] move some of the viz stuff (labels, etc) to the service definition/data-message to make more declarative (mix-in? where?)
 - [ ] allow widgets to publish/subscribe user selection events so multiple widgets can work together
 - [ ] login
```
"login" should be marked too

austinpaine replied:
done

cawasser commented on resources/sql/queries.sql:
```diff
@@ -46,3 +46,109 @@ DELETE FROM services WHERE id = :id;
+-- :result :raw
+-- :doc Create users table
+create table users (
+    username        VARCHAR(30) PRIMARY KEY,
```
make users table TEXT to be consistent

cawasser commented on src/clj/dashboard_clj/routes.clj:
```diff
@@ -23,15 +25,81 @@
+;;;;; Layout Saving Routes
+
+  ;TODO test removing the write-str here and put the de-string back in url handler
+  (GET "/layout" req
```
we need to look into how to document/comment routes. we also need to look into how we can start to split this file up across multiple namespaces by semantic purpose (users/login, layout, data subscriptions, etc.)

cawasser commented on src/cljs/vanilla/login_modal.cljs:
```diff
@@ -0,0 +1,211 @@
+         ;; This works and returns the handlers response
+         :format          :json
+         :params          credentials
+         :handler         #(prn "User created!")
```
(js/toastr.success "Welcome to Vanilla!")

cawasser commented on src/cljs/vanilla/login_modal.cljs:
```diff
@@ -0,0 +1,211 @@
+         :format          :json
+         :params          credentials
+         :handler         #(prn "User created!")
+         :on-error        #(prn "ERROR creating the user! " %)}))
```
(js/toastr.error "Can't create account. Try again.")

cawasser commented on src/cljs/vanilla/update_layout.cljs:
```diff
@@ -1,4 +1,17 @@
+        {:format          (ajax/json-request-format {:keywords? true})
+         :response-format (ajax/json-response-format {:keywords? true})
+         :params          {:widgets (clojure.core/pr-str layout)}       ;convert the whole layout struct to a string to preserve values
+         :handler         #(prn "Layout SAVED!")
```
(js/toastr.success "Layout saved.")

cawasser commented on src/cljs/vanilla/update_layout.cljs:
```diff
@@ -1,4 +1,17 @@
+         :on-error        #(prn "ERROR saving the layout " %)}))
 
 
 (defn- apply-updates [new-layout widget]
```
(js/toastr.error "Layout could not be saved.")

cawasser commented:
why is there a "user_db" separate from "vanilla_db"? we only want 1 database, with multiple tables in it.

cawasser commented on src/clj/vanilla/db/core.clj:
```diff
@@ -22,6 +22,10 @@
   "SQLite database connection spec."
   {:dbtype "sqlite" :dbname "vanilla_db"})
 
+(def users-db
```
don't think we want a second database file. could cause trouble in a deployed docker container, as we have to set up the single Database-URL arg for the app.

gpierce3 replied:
Agree with having only a single db for the app.
Does Chris's comment about "single Database-URL" mean we'll have problems accessing another system's db?

cawasser replied:
We need to look into how we can specify db config information to the app when it’s inside a deployed container. Some stuff may be “working” simply because thing running local still have access to local resources. 

But we don’t want OUR data for running/configuring Vanilla in multiple places. Reaching to other data sources is different

cawasser replied:
Scratch that. We didn’t use an external config for the database(s). We hard coded them into db.core. 

Let’s talk about the hardcoding, but I think we should still get down to 1 db for “internal” use

austinpaine replied:
Yeah that shouldnt be a problem, I think chad just had it already created when we merged together.  Should be quick to just make it a table in vanilla db, will do.

cawasser replied:
Remove 3 lines and change the few calls that use user_db

austinpaine replied:
Done and tested

chadarm replied:
Yeah just to comment on it, user_db was created for testing purposes. Thanks for taking care of it @austinpaine !

gpierce3 reviewed (commented):
Just curious, why the change from varchar to text in queries.sql?
Nothing further to add beyond what others have said.

austinpaine commented:
@gpierce3 Basically SQLite doesnt care at all about specific types, or the (length) whatsoever.  It will store whatever size data you want as a string no matter what, unless its numbers.

 More info: https://www.sqlite.org/datatype3.html

Commit:
changed user table to be in vanilla-db, updated readme

Commit:
removed user db

Commit:
added toastr libs and imports.  added it to layout save but theres a bug and it triggers both a success and fail message

Commit:
Toasts almost all working right.  Layout works, when a user sign up or login fails, that error isnt triggering the error toast

cawasser commented on src/cljs/vanilla/login_modal.cljs:
```diff
@@ -0,0 +1,216 @@
+         :response-format (ajax/json-request-format {:keywords? true})
+         :params          credentials
+         :handler         #(rf/dispatch [:login-message % "Account created!"])
+         :on-error        #(rf/dispatch [:login-message % "Failed to create account, please try again."])}))
```
:error-handler, not ":on-error"

cawasser commented on src/clj/dashboard_clj/routes.clj:
```diff
@@ -23,15 +25,79 @@
+
+  (GET "/verify-user" req
+    {:status 200
+    :headers {"Content-Type" "text/json; charset=utf-8"}
```
the "user" routes are not indented correct, so parinfer may cause problems down the road. please push the :headers and :body segments right so they line up with :status "inside" the {}-map

Commit:
commiting a clean vanilla_db, fixed error-handler for login post

Commit:
Fixed new-widget loading and saving to db, build-fn is removed before save and re-added from widget-defs on load

Commit:
fixed long login freeze.  Moved layout loading into update-layout ns to better organize functionality.  login modal now handles calling the layout load.  toasts function right, no "layout saved" on first login

austinpaine commented:
@cawasser @chadarm @gpierce3 @bfhub  I have completed all the changes requested, fixed the handling of new widgets, and added toasts for login and layout saving.  Please pull and retest/re-review changes so we can merge this in.

austinpaine commented:
Known bug! On new user sign up toast popup is green but says "failed to make user message".  However the user sign up **does succeed** and gets in the DB to make a new user.  Should be able to immediately login with new user.  The call is succeeding no problem but the error-handler is being triggered and I could not figure out why.  Took a break to get started on the swagger task for now

bfhub reviewed (commented):
verified handling the new widgets.

chadarm commented:
It looks great, saves my layout and can login to see it repopulate my layout. Awesome work, love the notification flags too.

Commit:
fixed weird json response resulting in failed post on new user creation

cawasser committed:
add some css (http://bulma.io) styling to the login-modal

cawasser reviewed (approved):
Nice work. This sets us up nicely for our next enhancements

austinpaine commented:
Tested Chris changes, works and looks great.
